Add --seed for reproducible test runs (and --chaos)

ysbaddaden opened the pull request:
Tests are now shuffled at runtime after initializing the PRNG with a seed (specified or random) so tests can be run in a reproducible order for debug purposes.

Bonus: tests are now shuffled globally, and workers will pick tests one by one, instead of picking a test suite and running all its tests.

closes #14

ysbaddaden committed:
Shuffle tests at runtime

Shuffle the tests in a suite so they're shuffled dynamically
at runtime, not statically at compile time.

ysbaddaden committed:
Collect and run tests globally (not per test suite)

Instead of collecting test suites, then shuffling tests in a test
suite, collect all tests for all test suites, then shuffle it all
for true chaos.

ysbaddaden committed:
Add --seed option for reproducible test runs

Prints the random seed used to shuffle the test suite, allowing to
reproduce test failures that stem from a certain execution order.

ysbaddaden commented:
@splattael does it work for you, now?

splattael commented:
@ysbaddaden Yes!! I've tried on https://github.com/splattael/random-failures and on another project at work and the test failures are now reproducible! :sparkles: 

Thank you :green_heart:

Sija commented on src/minitest.cr:
```diff
@@ -67,28 +98,38 @@ module Minitest
   def self.run(args = nil)
     process_args(args) if args
-    reporter.start
+    puts options
```
Debug leftover?

ysbaddaden replied:
No. Without this print, we wouldn't know the run options, especially the default seed; the filters are important, too; to know how to reproduce a random failure: just copy paste the options.

Sija commented on src/runnable.cr:
```diff
@@ -6,17 +6,31 @@ module Minitest
+
+    # Builds, at compile time, an Array with the test method name (String) and a
+    # proc to call that method. The Array will then be shuffled at runtime.
+    macro def self.collect_tests
```
`macro def` -> `def`

Sija commented:
Will it blend, err 🚢 ?

ysbaddaden commented:
Yes, it will. I'm just wondering whether shuffling every test case from any test suite was really a good idea. I mean, not even the original minitest does this (it shuffles suites, then shuffles test cases in each suite).

Oh well, with concurrent runners, shuffling everything means running the test suite in a real chaos, which is a good thing... I guess?

ysbaddaden committed:
Introduce --chaos argument to shuffle everything

Instead of always collecting all tests from all test suites then
shuffling and running all tests, introduce a `--chaos` parameter to
enable this feature on demand, along with --parallel. That is, on
demand absolute chaos.

Defaults back to run each test suite in whatever order, then
shuffling the tests for each suite. That is, less chaos.

ysbaddaden committed:
Reproducible tests: use custom Random instance

Use a custom Random instance for shuffling tests, instead of relying
on the global `rand` which may be influenced by the application at
runtime.

ysbaddaden commented:
Rebased on master. Reworked to introduce a `--chaos` parameter to selectively shuffle all tests from all test suites, or fallback to shuffle the suites, then shuffle the tests for each suite. Uses a custom `Random::DEFAULT` instance to avoid being influenced by potential runtime usages of `rand`.

Should be good to merge. Please test :)

splattael commented:
Wow, nice! Thank you :purple_heart: 

Sadly, I am stuck on Crystal 0.23.1 as Crystal 0.24.0 has not been released officially.

@ysbaddaden Is there a way to make minitest.cr backwards-compatible with Crystal 0.23.1?

ysbaddaden commented:
No, the breaking changes are complex to work around (especially the `macro def` change).

ysbaddaden commented:
Let's merge. If you find any bug on master, please open an issue :)
